flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -664,12 +699,96 @@ static std::string getIntrinsicWrapperName(const llvm::StringRef &intrinsic,
   return name;
 }
 
+/// fir::ExtendedValue to mlir::Value translation layer
+
+fir::ExtendedValue toExtendedValue(mlir::Value val,
+                                   Fortran::lower::FirOpBuilder &builder,
+                                   mlir::Location loc) {
+  assert(val && "optional unhandled here");
+  auto type = val.getType();
+  auto base = val;
+  mlir::Value len;
+  auto indexType = builder.getIndexType();
+  llvm::SmallVector<mlir::Value, 2> extents;
+
+  if (auto refType = type.dyn_cast<fir::ReferenceType>())
+    type = refType.getEleTy();
+
+  if (auto arrayType = type.dyn_cast<fir::SequenceType>()) {
+    type = arrayType.getEleTy();
+    llvm::ArrayRef<fir::SequenceType::Extent> shape = arrayType.getShape();
+    if (auto charType = type.dyn_cast<fir::CharacterType>()) {
+      auto cstLen = arrayType.getShape()[0];
+      shape = shape.drop_front();
+      if (cstLen == fir::SequenceType::getUnknownExtent())
+        mlir::emitError(loc, "No runtime length for given mlir::Value");
+      len = builder.createIntegerConstant(loc, indexType, cstLen);
+    }
+    for (auto extent : shape) {
+      if (extent == fir::SequenceType::getUnknownExtent())
+        extents.emplace_back(mlir::Value{});
+      else
+        extents.emplace_back(
+            builder.createIntegerConstant(loc, indexType, extent));
+    }
+  } else if (type.isa<fir::BoxCharType>()) {
+    std::tie(base, len) =
+        Fortran::lower::CharacterExprHelper{builder, loc}.createUnboxChar(val);
+  } else if (type.isa<fir::CharacterType>()) {
+    len = builder.createIntegerConstant(loc, indexType, 1);
+  } else if (type.isa<fir::BoxType>() || type.isa<fir::RecordType>()) {
+    mlir::emitError(loc, "descriptor or derived type not yet handled");
+  }
+
+  if (len && !extents.empty())
+    return fir::CharArrayBoxValue{base, len, extents};
+  if (len)
+    return fir::CharBoxValue{base, len};
+  if (!extents.empty())
+    return fir::ArrayBoxValue{base, extents};
+  return base;
+}
+
+llvm::SmallVector<fir::ExtendedValue, 2>
```

Comment:
Return a SmallVectorImpl<>?

Reply:
Thanks for pointing to this, it often annoyed me how to pass those as arguments, `SmallVectorImpl` definitely solve that. However, it seems it cannot be used as a returned value since `SmallVectorImpl` is the base class without the storage. Given I am not calling out the type when using this function, I think it is OK to stay like this.

Reply:
Oh, right. Yes, you'll find LLVM code usually declares the vector in the caller and returns the populated vector through an argument to the call. It keeps everything on the stack.

Reply:
OK, I was a bit too functional programming oriented here, I have removed those two functions that were not much used. It's cleaner to handle new array and the loop locally.